### ISSUE TITLE
Azure DevOps pipeline hangs on unit tests after updating from 2.6.6 to 2.7.0

### ISSUE THREAD

**Issue by ajeglin** [xunit/xunit#2904]
After updating from version 2.6.6 to 2.7.0, the unit test step in an Azure DevOps pipeline now hangs indefinitely until the pipeline times out. If the package is downgraded back to 2.6.6, it works as expected.

The final message that gets written to the console before it hangs is:
_Starting test execution, please wait...
A total of 1 test files matched the specified pattern._

Here are all of the package references in the unit tests project:
```
<ItemGroup>
  <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.9.0" />
  <PackageReference Include="xunit" Version="2.7.0" />
  <PackageReference Include="xunit.runner.visualstudio" Version="2.5.7">
    <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    <PrivateAssets>all</PrivateAssets>
  </PackageReference>
  <PackageReference Include="coverlet.collector" Version="6.0.2">
    <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
    <PrivateAssets>all</PrivateAssets>
  </PackageReference>
  <PackageReference Include="Moq" Version="4.20.70" />
</ItemGroup>
```

The unit tests step of the YAML pipeline is being run using the `DotNetCoreCLI@2` task.

This issue has been seen on both the windows-latest and ubuntu-latest built-in agent specifications.

Tests run as expected on version 2.7.0 from Visual Studio - this issue only affects the pipeline. It has been identified in multiple projects.

**Comment by bradwilson** [comment]
What does the task that hangs look like?

Do you have a custom `xunit.runner.json`?

**Comment by bradwilson** [comment]
I have a project set up for another purpose (it only has 2 tests in it) with the following YAML:

```yaml
trigger:
- main

pool:
  vmImage: windows-latest

steps:
- task: DotNetCoreCLI@2
  inputs:
    command: 'build'

- task: DotNetCoreCLI@2
  inputs:
    command: 'test'
    arguments: '-v:normal'
```

Upgrading it to 2.7.0 does not cause any runtime failures:

![image](https://github.com/xunit/xunit/assets/16944/039d1e36-21a4-4cf9-a503-c7deed130ce2)

**Comment by ajeglin** [comment]
No, I don't have a custom `xunit.runner.json`.

I simplified my YAML in the pipeline to match yours as above and the issue still occurs. I have done some more investigating and identified a handful of tests that are causing it (even though they run and pass locally in Visual Studio).

I am out of the office tomorrow but I will look to put a sample app together on Wednesday that shows the problem. I think it may related to mocking HTTP requests,

**Comment by bradwilson** [comment]
One typical major difference between your local machine and a CI build is the number of available cores. If you have code that can hold onto a worker thread indefinitely, that might be causing a deadlock in the worker pool.

An easy test for this would be to set [`maxParallelThreads`](https://xunit.net/docs/configuration-files#maxParallelThreads) to `-1` and see if the problem disappears.

**Comment by bradwilson** [comment]
It might be easier to do [via command line](https://xunit.net/docs/runsettings) in your case rather than adding a `xunit.runner.json` file.

**Comment by ajeglin** [comment]
If I set the `maxParallelThreads` setting I then get a test failure, rather than it hanging. It fails because I have Moq configured in Strict mode and it seems to think I haven't set up a mocked method:
```
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. Moq.MockException: IDataRepository.GetItems<ProjectSettings>(Func<IQueryable<ProjectSettings>, IQueryable<ProjectSettings>>) invocation failed with mock behavior Strict.
All invocations on the mock must have a corresponding setup.
```
The Mock instance in question is instantiated in the constructor of the test class, as normal. The tests pass locally if I use the built in Visual Studio test runner, but are all inconclusive if I run the whole test suite using the ReSharper test runner.

I am struggling to re-create the original problem in a standalone demo project, however. If I reduce the number of tests in the test project, my pipeline works on version 2.7.0. If I re-instate all of the tests, it hangs again.

**Comment by bradwilson** [comment]
That sounds more serious, like there is a corrupted stack somewhere.

**Comment by bradwilson** [comment]
That said, if you can provide a stack trace, I may be able to figure out where this is being thrown and see if we can catch it so that it doesn't crash the test process.

**Comment by ajeglin** [comment]
Sure, here is the stack trace. It doesn't look like any XUnit classes are listed in it, though...

This is also an issue that only occurs on version 2.7.0. If I rollback to 2.6.6 the tests all run and pass with `maxParallelThreads `set to -1.

```
The active test run was aborted. Reason: Test host process crashed : Unhandled exception. Unhandled exception. Moq.MockException: IDataRepository.GetItems<ProjectSettings>(Func<IQueryable<ProjectSettings>, IQueryable<ProjectSettings>>) invocation failed with mock behavior Strict.
All invocations on the mock must have a corresponding setup.
   at Moq.FailForStrictMock.Handle(Invocation invocation, Mock mock) in /_/src/Moq/Interception/InterceptionAspects.cs:line 434
   at Moq.Mock.Moq.IInterceptor.Intercept(Invocation invocation) in /_/src/Moq/Interception/Mock.cs:line 27
   at Moq.CastleProxyFactory.Interceptor.Intercept(IInvocation underlying) in /_/src/Moq/Interception/CastleProxyFactory.cs:line 218
   at Castle.DynamicProxy.AbstractInvocation.Proceed()
   at Castle.Proxies.IDataRepositoryProxy.GetItems[TItem](Func`2 queryableFunc)
   at AdoMetrics.Services.DatabaseConfigUpdateService.DeleteExistingSettings(IEnumerable`1 projectConfigs) in /home/vsts/work/1/s/src/AdoMetrics/AdoMetrics.Services/DatabaseConfigUpdateService.cs:line 122
   at AdoMetrics.Services.DatabaseConfigUpdateService.<>c__DisplayClass5_0.<<UpdateProjectConfigAsync>b__0>d.MoveNext() in /home/vsts/work/1/s/src/AdoMetrics/AdoMetrics.Services/DatabaseConfigUpdateService.cs:line 35
--- End of stack trace from previous location ---
   at AdoMetrics.UnitTests.Services.DatabaseConfigUpdateServiceTests.<>c.<<SetupResilientTransactionMock>b__5_1>d.MoveNext() in /home/vsts/work/1/s/src/AdoMetrics/AdoMetrics.UnitTests/Services/DatabaseConfigUpdateServiceTests.cs:line 37
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
Moq.MockException: IDataRepository.GetItems<ProjectSettings>(Func<IQueryable<ProjectSettings>, IQueryable<ProjectSettings>>) invocation failed with mock behavior Strict.
All invocations on the mock must have a corresponding setup.
   at Moq.FailForStrictMock.Handle(Invocation invocation, Mock mock) in /_/src/Moq/Interception/InterceptionAspects.cs:line 434
   at Moq.Mock.Moq.IInterceptor.Intercept(Invocation invocation) in /_/src/Moq/Interception/Mock.cs:line 27
   at Moq.CastleProxyFactory.Interceptor.Intercept(IInvocation underlying) in /_/src/Moq/Interception/CastleProxyFactory.cs:line 218
   at Castle.DynamicProxy.AbstractInvocation.Proceed()
   at Castle.Proxies.IDataRepositoryProxy.GetItems[TItem](Func`2 queryableFunc)
   at AdoMetrics.Services.DatabaseConfigUpdateService.DeleteExistingSettings(IEnumerable`1 projectConfigs) in /home/vsts/work/1/s/src/AdoMetrics/AdoMetrics.Services/DatabaseConfigUpdateService.cs:line 122
   at AdoMetrics.Services.DatabaseConfigUpdateService.<>c__DisplayClass5_0.<<UpdateProjectConfigAsync>b__0>d.MoveNext() in /home/vsts/work/1/s/src/AdoMetrics/AdoMetrics.Services/DatabaseConfigUpdateService.cs:line 35
--- End of stack trace from previous location ---
   at AdoMetrics.UnitTests.Services.DatabaseConfigUpdateServiceTests.<>c.<<SetupResilientTransactionMock>b__5_1>d.MoveNext() in /home/vsts/work/1/s/src/AdoMetrics/AdoMetrics.UnitTests/Services/DatabaseConfigUpdateServiceTests.cs:line 37
--- End of stack trace from previous location ---
   at System.Threading.Tasks.Task.<>c.<ThrowAsync>b__128_1(Object state)
   at System.Threading.ThreadPoolWorkQueue.Dispatch()
   at System.Threading.PortableThreadPool.WorkerThread.WorkerThreadStart()
```

**Comment by bradwilson** [comment]
Bummer, with nothing of us in the stack trace, there's not much for me to go on. I did a quick peek through the changes between 2.6.6 and 2.7.0 and most of it was assertion changes. The only big core change that I could see was related to opting out of the AsyncTestSyncContext for async tests which return `Task`.

Hopefully you'll be able to provide a repro project.

In the meantime, as a test to see if this is the problematic issue, you can try to reproduce the issue with the `2.6.7-pre.10` (the build before the feature was introduced) vs. `2.6.7-pre.11` (the build where the feature was added). If the code runs with 10 but fails with 11, then it's the `AsyncTestSyncContext` removal that's causing the problem.

Information on using our CI builds: https://xunit.net/docs/using-ci-builds

Here is the full list of interim builds between 2.6.6 and 2.7.0:

- `2.6.7-pre.1` is the version renumber (no new code)
- `2.6.7-pre.3` adds assertion support for `IAsyncEnumerable<T>`
- `2.6.7-pre.4` assertion library update (no new code)
- `2.6.7-pre.5` uses `Enum.ToObject` instead of custom conversion logic
- `2.6.7-pre.6` re-introduces a deleted constructor to `SynchronousMessageBus`
- `2.6.7-pre.7` emits warning messages for broken configuration files
- `2.6.7-pre.8` adds additional stack trace information to `Assert.Collection`
- `2.6.7-pre.10` update to XML documentation (no new code)
- `2.6.7-pre.11` adds opting out of `AsyncTestSyncContext`
- `2.6.7-pre.12` fast fail if a test uses `Timeout` but isn't async
- `2.6.7-pre.14` updating build system (no new code)
- `2.6.7-pre.15` updating build system (no new code)
- `2.6.7-pre.16` updates `Assert.Contains` and `Assert.DoesNotContain` with expanded handling for sets

**Comment by BEagle1984** [comment]
I'm having the same issue, even on local. With 2.6.6 is all fine, but some tests are hanging after the upgrade to 2.7.0.
Interestingly, for some reason, the issue seems to be worse with .NET 6 than .NET 8 (I'm multitargeting the same solution and I test the same codebase against both framework versions).
My library makes heavy use of asynchronous tasks.

I will look deeper into this and let you know if I figure out something helpful.

My project: https://github.com/BEagle1984/silverback/tree/release/5.0.0

**Comment by bradwilson** [comment]
@BEagle1984 Which project(s) are hanging?

**Comment by bradwilson** [comment]
@BEagle1984 I find your settings puzzling:

```
  "parallelizeTestCollections": false,
  "parallelizeAssembly": false,
  "maxParallelThreads": 10_000_000_000,
```

If parallelization is off, what is the point of requesting 10 billion threads? Why not use -1 ("unlimited") instead?

Also, you should be aware that `parallelizeAssembly` is ignored by VSTest (`dotnet test`, `dotnet vstest`, Test Explorer, etc.). When I just run `dotnet test` at the top level, of if I run every project in Visual Studio, it runs the assemblies in parallel.

**Comment by bradwilson** [comment]
@BEagle1984 Comparing 2.6.6 and 2.7.0, I was able to reproduce the hang in `Silverback.Tests.Integration.E2E.Kafka.RebalanceFixture.Rebalance_ShouldCommitPendingOffsets`. The test actually runs to completion and the hang is during your cleanup code.

Here's the task list, sorted by name:

![image](https://github.com/xunit/xunit/assets/16944/38ba09e2-0959-4b49-bc4f-c151614ae0ec)

I believe I've located your deadlock. This is the code for `OnApplicationStopped`:

```csharp
[SuppressMessage("Usage", "VSTHRD002:Avoid problematic synchronous waits", Justification = "Reviewed")]
private void OnApplicationStopped() => _brokerDisconnectedTaskCompletionSource.Task.Wait();
```

This uses `.Wait()`, which is what's causing the deadlock.

Why is this surfacing in 2.7.0 but not in 2.6.6? I honestly don't know yet. I'm still looking into that.

**Comment by bradwilson** [comment]
I can verify that changing the test from `async Task` to `async void` allows the test to run to completion, so the deadlock in question is definitely being hidden by [`AsyncTestSyncContext`](https://github.com/xunit/xunit/blob/v2/src/xunit.execution/Sdk/AsyncTestSyncContext.cs).

**Comment by bradwilson** [comment]
@BEagle1984 I'm able to get all your tests to pass on 2.7.0 by removing all the potential deadlocks.

I added this to `Silverback.Util.TaskExtensions`:

```csharp
[SuppressMessage("Usage", "VSTHRD002:Avoid problematic synchronous waits", Justification = "Uses SpinWait to prevent deadlocks")]
public static void SpinWait(this Task task, CancellationToken token = default)
{
    SpinWait spin = default;

    while (!task.IsCompleted)
    {
        spin.SpinOnce();
        token.ThrowIfCancellationRequested();
    }

    task.GetAwaiter().GetResult();
}

[SuppressMessage("Usage", "VSTHRD002:Avoid problematic synchronous waits", Justification = "Uses SpinWait to prevent deadlocks")]
public static T SpinWait<T>(this Task<T> task, CancellationToken token = default)
{
    SpinWait spin = default;

    while (!task.IsCompleted)
    {
        spin.SpinOnce();
        token.ThrowIfCancellationRequested();
    }

    return task.GetAwaiter().GetResult();
}

[SuppressMessage("Usage", "VSTHRD002:Avoid problematic synchronous waits", Justification = "Uses SpinWait to prevent deadlocks")]
public static void SpinWait(this ValueTask task, CancellationToken token = default)
{
    SpinWait spin = default;

    while (!task.IsCompleted)
    {
        spin.SpinOnce();
        token.ThrowIfCancellationRequested();
    }

    task.GetAwaiter().GetResult();
}

[SuppressMessage("Usage", "VSTHRD002:Avoid problematic synchronous waits", Justification = "Uses SpinWait to prevent deadlocks")]
public static T SpinWait<T>(this ValueTask<T> task, CancellationToken token = default)
{
    SpinWait spin = default;

    while (!task.IsCompleted)
    {
        spin.SpinOnce();
        token.ThrowIfCancellationRequested();
    }

    return task.GetAwaiter().GetResult();
}
```

I searched your source for all instances of `VSTHRD002` and replaced all the calls to `.Wait()`, `.Result`, or `.GetAwaiter().GetResult()` with `.SpinWait()`. No more hangs.

I've attached a full patch for you to review: [deadlocks.patch](https://github.com/xunit/xunit/files/14830855/deadlocks.patch)

You already had had `AsyncHelper.RunSynchronously` which spawned off a new task which I also replaced with `.SpinWait()`, though you could argue that they're not exactly the same so that one change is not strictly necessary. Leaving the old implementation in place still allows all the tests to pass, as that was not a source of a deadlock (just a different way to accomplish what `.SpinLock()` does). In point of fact, you could also use your existing implementation instead of the `SpinLock` class to remove the deadlocks.

**Comment by bradwilson** [comment]
@BEagle1984 In order to prevent deadlocks, it's my opinion that you should never be directly calling `.Wait()`, `.Result`, or `.GetAwaiter().GetResult()` until you know the task is complete. For libraries used by third parties, you don't have enough control over the execution environment to judge such calls as being safe. Adapting such async internals to be synchronous safely requires a technique like your `AsyncHelper.RunSynchronously` or my `SpinLock`.

I would strongly recommend you patch your code to prevent these deadlocks. It honestly could be sheer luck that none of your consumers have hit any of the potential deadlocks yet.

**Comment by BEagle1984** [comment]
The library has been used for many years and we don't experience any deadlock in production (there were some occasional bugs but got fixed).
Plus, the tests also worked fine with version 2.6.6 of xunit. Something changed between 2.6.6 and 2.7.0, which caused the issue. Maybe it just highlights a latent problem, but something must have changed. I can switch back and forth between 2.6.6. and 2.7.0 and the newer version always hang on the same test (after it is finished executing actually).

I don't think I'm calling `.Wait()` or `.Result` in an unsafe manner but I might have overlooked something. Did you spot a specific misusage?

**Comment by BEagle1984** [comment]
> @BEagle1984 I find your settings puzzling:
> 
> ```
>   "parallelizeTestCollections": false,
>   "parallelizeAssembly": false,
>   "maxParallelThreads": 10000000000,
> ```
> 
> If parallelization is off, what is the point of requesting 10 billion threads? Why not use -1 ("unlimited") instead?

At some point, I had them hanging on the azure devops pipeline and I made some tests which I forgot to cleanup.
 
> Also, you should be aware that `parallelizeAssembly` is ignored by VSTest (`dotnet test`, `dotnet vstest`, Test Explorer, etc.). When I just run `dotnet test` at the top level, of if I run every project in Visual Studio, it runs the assemblies in parallel.

This is weird though, isn't `false` the default? I expected them to _not_ run in parallel if the setting is ignored.

**Comment by BEagle1984** [comment]
@bradwilson unfortunately the same test still hangs on my machine after having applied your patch and upgraded again to 2.7.0.

![image](https://github.com/xunit/xunit/assets/30572406/b11e7fde-9445-47bd-afdb-86076c89aae5)

**Comment by bradwilson** [comment]
@BEagle1984 

> This is weird though, isn't false the default? I expected them to not run in parallel if the setting is ignored.

There isn't just one default (and unfortunately, everything here ignores the `parallelizeAssembly` setting in `xunit.runner.json`).

- `dotnet test` runs in parallel by default. AFAIK the only way to disable this behavior is to run your test assemblies individually one by one rather than running `dotnet test` against the whole solution.
-  Test Explorer runs in parallel by default. There used to be a menu option to enable/disable parallelism but I don't see it any more in VS 2022.
- `dotnet vstest` runs sequentially by default. You can use the `-Parallel` switch with it to enable parallelism.

> Yeah, something must definitely have changed.

We have two synchronization contexts in xUnit.net.

There is one for limiting parallelism when you have enabled test collection parallelism. When it's disabled, the sync context isn't there. That has been like that since at least 2.0 was released, and that behavior hasn't changed.

There is one for supporting `async void` test methods. Prior to 2.7.0, this sync context was always in place; starting with 2.7.0, it's only in place when we believe we have identified the test method is `async void`.

So what's different is that previously you had a sync context, and now you don't. When there's no underlying sync context (which there wouldn't be), `AsyncTestSyncContext` on .NET Core will use `TaskFactory.StartNew` with the default scheduler to do sync context continuations, which should push the work back onto the same thread pool that .NET uses by default. So the presence or absence of the `AsyncTestSyncContext` should be the same, which is why it's so curious why the behavior has changed for you.

> the same test still hangs on my machine

What test runner are you using? That's not Test Explorer.

**Comment by BEagle1984** [comment]
I'm running them in Rider, which I thought it would simply do a `dotnet test`.

**Comment by bradwilson** [comment]
Rider definitely has their own custom test runner, which they support (not us). It would be interesting to see if you can reproduce the problem in `dotnet test` or Test Explorer.

**Comment by bradwilson** [comment]
I'll dig in some more and see if I can find anything else.

**Comment by BEagle1984** [comment]
I can definitely reproduce the same behavior in VS via the Test Explorer.

![image](https://github.com/xunit/xunit/assets/30572406/25d7d66f-cebe-45f1-9a54-0a5cc8e3f3fc)

It just hangs on the same test, with your patch freshly applied, and after upgrading the xunit package to 2.7.0.

**Comment by BEagle1984** [comment]
Is there anything I can do to provide you with more data?

**Comment by bradwilson** [comment]
There is a workaround for you: convert your async tests from `async Task` to `async void`. This will ensure that `AsyncTestSyncContext` continues to be used.

I'm still of the belief that there is a fundamental deadlock that I haven't uncovered in your code yet, but I'm not sure how much more time I can spend trying to track it down. It's also possible that the spin lock is contributing to the deadlocks in a different way, since that doesn't yield the thread (unlike your `AsyncHelper.RunSynchronously` implementation).

**Comment by BEagle1984** [comment]
I don't think that the SpinLock is causing any harm. I tried using the `AsyncHelper` everywhere (which I prefer anyway) but it's not any better.
Thank you very much for your help so far. :hearth:
I will, of course, keep investigating and let you know if I discover anything.

**Comment by ajeglin** [comment]
> Bummer, with nothing of us in the stack trace, there's not much for me to go on. I did a quick peek through the changes between 2.6.6 and 2.7.0 and most of it was assertion changes. The only big core change that I could see was related to opting out of the AsyncTestSyncContext for async tests which return `Task`.
> 
> Hopefully you'll be able to provide a repro project.
> 
> In the meantime, as a test to see if this is the problematic issue, you can try to reproduce the issue with the `2.6.7-pre.10` (the build before the feature was introduced) vs. `2.6.7-pre.11` (the build where the feature was added). If the code runs with 10 but fails with 11, then it's the `AsyncTestSyncContext` removal that's causing the problem.
> 
> Information on using our CI builds: https://xunit.net/docs/using-ci-builds
> 
> Here is the full list of interim builds between 2.6.6 and 2.7.0:
> 
> * `2.6.7-pre.1` is the version renumber (no new code)
> * `2.6.7-pre.3` adds assertion support for `IAsyncEnumerable<T>`
> * `2.6.7-pre.4` assertion library update (no new code)
> * `2.6.7-pre.5` uses `Enum.ToObject` instead of custom conversion logic
> * `2.6.7-pre.6` re-introduces a deleted constructor to `SynchronousMessageBus`
> * `2.6.7-pre.7` emits warning messages for broken configuration files
> * `2.6.7-pre.8` adds additional stack trace information to `Assert.Collection`
> * `2.6.7-pre.10` update to XML documentation (no new code)
> * `2.6.7-pre.11` adds opting out of `AsyncTestSyncContext`
> * `2.6.7-pre.12` fast fail if a test uses `Timeout` but isn't async
> * `2.6.7-pre.14` updating build system (no new code)
> * `2.6.7-pre.15` updating build system (no new code)
> * `2.6.7-pre.16` updates `Assert.Contains` and `Assert.DoesNotContain` with expanded handling for sets

Sorry for the delay getting back to you.

I can confirm that the issue appears in version `2.6.7-pre.11` and not in `2.6.7-pre.10`.

**Comment by bradwilson** [comment]
@ajeglin I'm going to guess that the issue is a deadlock for you as well, and the same workaround would apply: use `async void` rather than `async Task` for your test methods that are hung.

**Comment by bradwilson** [comment]
@BEagle1984 I did have a thought about your AsyncHelper implementation.

We use `TaskCreationOptions.LongRunning` in our continuation: https://github.com/xunit/xunit/blob/57af1d9f0b18181631cf6aa3d40a9f950468b1bc/src/common/XunitWorkerThread.cs#L31-L48

The [documentation](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskcreationoptions?view=net-8.0) describes it as such:

> Specifies that a task will be a long-running, coarse-grained operation involving fewer, larger components than fine-grained systems. It provides a hint to the [TaskScheduler](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.taskscheduler?view=net-8.0) that oversubscription may be warranted. Oversubscription lets you create more threads than the available number of hardware threads. It also provides a hint to the task scheduler that an additional thread might be required for the task so that it does not block the forward progress of other threads or work items on the local thread-pool queue.

I wonder if you updated your implementation if that would solve the problem for you.

**Comment by bradwilson** [comment]
@ajeglin I really have no idea why Moq has corrupted memory. That's not something I can address.

**Comment by ajeglin** [comment]
@bradwilson fair enough re the Moq exception, but the exception only occurs on the newer version of xunit and only when I used `async void` rather than `async Task`. There is nothing special or weird being done with Moq within the test. So the underlying cause of the exception still lies within xunit, not Moq.

**Comment by BEagle1984** [comment]
@bradwilson I'm not sure that ithose tasks are really "long running". I thought that was meant for the tasks that really do long-running synchronous processing for some reason and it only make sense if you don't spawn other tasks from within it.

In any case, there is indeed some sort of deadlock around the disconnect/cleanup code in Silverback. For some reason is only evident after the migration to 2.7.0 and only in that test. If I remove the `ApplicationStopped` registration, the disconnect and cleanup is executed when the objects are being disposed and I run into the same deadlock.
I haven't been able to properly understand where it's coming from and to fix it, but adding a `Task.Delay` at the end of the test method solves it, so I'm pretty sure it's due to a race condition that happens only in that particular case.

Sorry for having wasted your time! 😞

**Comment by mconnew** [comment]
If it's any help, I have been seeing test failures with CoreWCF tests running in CI when using 2.7.0 that I don't see when using 2.6.6. I think I have an idea what's causing it. The call to `AsyncTestSyncContext.WaitForCompletionAsync` won't complete until all queued delegates have completed execution. If a test launches a background Task, that could by queuing work to the AsyncTestSyncContext. The completion of the delegate depends on the thread pool having executed it (as you use ThreadPool.QueueUserWorkItem). When there's a lot of thread contention (as is the case in Azure DevOps as the build agents only have 2 cores), a lingering Task continuation delegate waiting to be executed will delay the completion of the test as the operation count needs to get to zero. When you removed always using `AsyncTestSyncContext`, tests now complete when the Task returned by the test method completes, regardless if there's any background Task's which are still going. When shutting down/cleaning up something (such as WebHost that CoreWCF uses), these background Task's get stopped by calling `CancellationToken.Cancel`. This tends to be a cooperative mechanism so sometimes can take longer to execute than the cleanup call which triggers it (might require a Task to be woken up such as setting a `AsyncManualResetEvent` which requires scheduling to the thread pool). By no longer waiting for these background Task's to complete before completing the test and starting the next one, you have increased the contention on the thread pool. The test failures that CoreWCF is seeing are caused by `TimeoutException` getting thrown. In the specific sets of tests we're seeing fail, the timeout is happening because `HttpClient` continuations aren't executing in a timely manner and the request is timing out. A simple request/reply call which should take less than 1ms isn't completing within a minute. Some of the problems will be continuations having delayed execution on the CoreWCF receive/dispatch side too.  

Just in case it's brought up, I understand that an async method could be awaiting on something still needing to be executed and `AsyncTestSyncContext` could get to zero operations without a background Task completing. In the situation I described, the cancellation of a background Task is likely to have posted a callback to the sync context before the test ends.  

I'm working on implementing my own test method runner with my own sync context with some extra bookkeeping and will examine what other things are running in the background. I've had a really hard time replicating locally as you need 2 CPU cores to hit the issue. I've resorted to using remote debugging to try to analyze things. Running the tests on Linux via docker or wsl2 reproduces the issue until I run the tests with a debugger attached. When I do that, after about 30 tests have executed, things look to hang and no more tests complete, and at some point the remote dotnet processes exit and VS is left hanging waiting for the tests to complete seemingly unaware that nothing is still running on the remote OS. When remote debugging to a Windows VM with 2 CPU's, I had to use sysinternals CPU Stress to eat CPU to reproduce the issue. I found some things I could do to improve things, such as I discovered that `WebHost.CreateDefaultBuilder` creates a background thread for console logging by default, per test. So I switched away from that to configure everything explicitly and that improved things so I got less failures. I don't want to artificially limit the parallelism by picking a number to set for MaxParallelThreads, that seems like a band aid.

**Comment by bradwilson** [comment]
@mconnew One thing that's not clear to me yet is that I see a common pattern here of people hitting this problem has having _no_ sync contexts because they've disabled test collection parallelization (which means we won't use `MaxConcurrencySyncContext`). Our usage of `ThreadPool.QueueUserWorkItem` from `AsyncTestSyncContext` should be using the exact same thread pool that the `Task` system uses by default, so I don't yet understand how that changes the way things are scheduled.

Are you also disabling test collection parallelization?

**Comment by bradwilson** [comment]
> I'm working on implementing my own test method runner with my own sync context with some extra bookkeeping and will examine what other things are running in the background.

This seems like more work than just using `async void` instead of `async Task`, but that's for you to decide. You'd probably also want a Roslyn analyzer to force your developers to use `async void` so maybe not, though we could provide one out of the box that's disabled by default if that would help.

**Comment by bradwilson** [comment]
I just thought of another possibility, and would be interested to hear feedback:

- Re-enable `AsyncTestSyncContext` in v2
- Disable support for `async void` in v3 (we can detect this and just fast fail the test)

This would not solve the problem if/when you moved to v3, because disabling support would mean outright deleting the sync context, but it would at least get you back to running on v2 without the need to do any additional work.

**Comment by bradwilson** [comment]
Alright I thought of another option that's similar: create a configuration option to opt out of the `AsyncTestSyncContext`, and the people who care can do that. Setting that option will then fast fail any tests that are `async void`.

**Comment by mconnew** [comment]
@bradwilson, we apply `[assembly: Xunit.CollectionBehavior(MaxParallelThreads = -1)]` to our test assembly. The difference using `AsyncTestSyncContext` is you wait for the operation count to get to zero by calling `AsyncTestSyncContext.WaitForCompletionAsync` before completing the test. When you don't use the sync context, you complete the test when the Task the method returned completes. The only thing which could cause the failures I'm seeing is greater thread pool contention or more threads running. Removing the sync context can effectively increase the test concurrency if there's lingering work after the methods Task has completed.

**Comment by bradwilson** [comment]
Yeah, MaxParallelThreads of -1 will also remove the other sync context, so you too are in the "there are no sync contexts" situation.

**Comment by bradwilson** [comment]
> When you don't use the sync context, you complete the test when the Task the method returned completes.

I believe this to be the correct behavior. Do you disagree? That's fundamental to how `async`/`await` is supposed to work.

This would at least explain why I couldn't reason about this being a scheduling problem, because it's not a scheduling problem but rather a background task that we're unintentionally waiting for that normally nobody would've waited for.

**Comment by mconnew** [comment]
The old behavior had a nice side effect that during early in the development of CoreWCF, I had a couple of bugs where background Task's were left running and Xunit caught them because the test wouldn't complete and just hung.  

I think it's a philosophical question with no right answer. You could argue that a test isn't complete until all execution related to the test has ceased, but you could also argue that the test itself shouldn't complete its returned Task until all execution has ceased. The area where this becomes tricky is I don't believe the lingering work is coming from my code, I believe it's coming from .NET, either in asp.net core, or more likely HttpClient (we don't see this issue with our non-http tests which still uses asp.net core).  

I don't think the problem is that you are taking the Task completion as the signal to move along, the problem is that you weren't before, and now you are. The change in philosophy is what's tripping people up. I think whatever the behavior is in a major release should continue to be the case for that major release, but it's fine to switch between major release, ie v3 doing something different is fine, but 2.6 to 2.7 having a change in behavior is a problem.  

I'm going to persevere as I want to understand what's still lingering around that's causing problems as I see that as a potential bug I need to investigate. Now that you support `BeforeAfterTestAttribute` at the assembly level, anyone can easily put a sync context back with minimal effort, so I see it more of an annoyance than a blocker.

**Comment by bradwilson** [comment]
> you could also argue that the test itself shouldn't complete its returned Task until all execution has ceased

Yeah, this is where I land. It's not our responsibility to track down the things you left hanging around. If you chose to launch them into the background, that's your responsibility. The fact that we did this accidentally is a problem now, because as you can see people are popping up saying "why did you break us?!". 😂 

> 2.6 to 2.7 having a change in behavior is a problem

It was unintentional, to be sure, which is why I might end up re-enabling the sync context and giving people a way to opt out.

> Now that you support `BeforeAfterTestAttribute` at the assembly level, anyone can easily put a sync context back with minimal effort, so I see it more of an annoyance than a blocker.

I don't know if that's gonna work or not. Someone's going to have to test it. Pushing in a sync context generally only affects things that you call. It's not (currently) async so it might be able to work, but I can't honestly guarantee it. There's a lot of async stuff that happens internally in xUnit.net, so if that call ends up down an async path that is returned out of, rather than used to actually run the test, then the sync context is gonna disappear.

**Comment by mconnew** [comment]
I think you're right, I've been elbow deep in the implementation and the BeforeAfter stuff is called via what's effectively a Task.Run so I don't think the execution context will flow to the to the test execution.

**Comment by mconnew** [comment]
@bradwilson, I have some info for you about trailing delegates to be executed. It's not caused by anything in the test itself, there's an extra delegate posted to the sync context by the Task async machinations. Its job is to send an ETW event for TPL and is scheduled to run after a generated Task which backs an async method completes. When you have a high CPU/thread contention test running environment, this will act like a self throttling mechanism. The `AsyncTestSyncContext.WaitForCompletionAsync` method will take longer to complete the more contention there is due to this extra delegate thrown in by Task's. By not waiting for this extra delegate, you no longer have a self throttling mechanism and over schedule concurrently executing tests.

**Comment by bradwilson** [comment]
It's not clear from your analysis whether you think this is our fault and/or what we should do about it. 😂

**Comment by bradwilson** [comment]
The more I think about this, it's crazy to imagine that the design of the TPL is to add an extra task on top of every task, and not think that they'll eventually exhaust the thread pool if they don't get good/lucky scheduling.

**Comment by mconnew** [comment]
It's not a new Task, it's an extra continuation (which is just a delegate) after the Task completes. It's the equivalent of doing something like this:
```c#
public Task RunTestAsync()
{
    Task testTask = RunTestImplAsync();
    testTask.ContinueWith(state => { /* Emit a ETW event to track TPL */ }, null);
    return testTask;
}
```

You are waiting for testTask to complete, and as soon as it does, you move on to the next test. Meanwhile, the TPL infrastructure has scheduled that extra continuation. It's done with less overhead than ContinueWith creates as that creates another new Task and you would end up with recursion, but hopefully you get the general idea. The mechanism they are doing just replaces the regular continuation registered against the Task with a new one which calls the original continuation, then after that's executed, runs the new one which schedules this ETW continuation to the same SynchronizationContext. Because you previously had a custom sync context which you would wait on all pending continuations to be completed, you would wait for this extra continuation to run before moving on to the next test.  

But there's something else going on too that I'm trying to understand. I set MaxParallelThreads to 2, then added some bookkeeping code to my custom TestInvoker to keep track of the number of concurrent tests being executed by xunit. I discovered that xunit was starting as many as 60 tests concurrently. I'm going to create a base repro outside of my current project to understand what's going on there.

**Comment by bradwilson** [comment]
> But there's something else going on too that I'm trying to understand. I set MaxParallelThreads to 2, then added some bookkeeping code to my custom TestInvoker to keep track of the number of concurrent tests being executed by xunit. I discovered that xunit was starting as many as 60 tests concurrently.

You will typically see one test per test collection scheduled to run at any given time (and as each test collection runs out of tests, the number should go down). The [sync context](https://github.com/xunit/xunit/blob/v2/src/xunit.execution/Sdk/MaxConcurrencySyncContext.cs) keeps a list of the available task callbacks to be run, and the threads in the thread pool pull individual items from the list. (I notice the historical comment on code mentioning DNX, which is from back when .NET Core was called Project K. I assume the bug still exists, but there's really not much point in removing the code now.)

So this is pretty much as I'd expect, assuming you have ~ 60 test collections (the default behavior of which being 1 test collection per test class).

**Comment by bradwilson** [comment]
The base test assembly runner will run the collections sequentially:

https://github.com/xunit/xunit/blob/6d6b413484e948bffad27efb8326113bf90a7d7f/src/xunit.execution/Sdk/Frameworks/Runners/TestAssemblyRunner.cs#L241-L246

This is overridden in the `XunitTestAssemblyRunner` to run the test collections in parallel, as well as supporting the notion that you can mark a test collection to not be run in parallel (so we segregate the parallel ones and run them all in parallel, and then after they're all done running, we run the non-parallel ones in sequence):

https://github.com/xunit/xunit/blob/6d6b413484e948bffad27efb8326113bf90a7d7f/src/xunit.execution/Sdk/Frameworks/Runners/XunitTestAssemblyRunner.cs#L209-L250

The base `TestCollectionRunner` will run the classes sequentially (and we don't override this):

https://github.com/xunit/xunit/blob/6d6b413484e948bffad27efb8326113bf90a7d7f/src/xunit.execution/Sdk/Frameworks/Runners/TestCollectionRunner.cs#L128-L133

The base `TestClassRunner` will run the methods sequentially (and we don't override this):

https://github.com/xunit/xunit/blob/6d6b413484e948bffad27efb8326113bf90a7d7f/src/xunit.execution/Sdk/Frameworks/Runners/TestClassRunner.cs#L226-L231

This continues on further down the stack (to test cases, then tests). The overrides are available so third parties can use much of our execution machinery but tweak things like being able run classes in the same test collection in parallel (and we will likely enable some of this behavior in v3 as it's long been a request to give users more control over parallelism without forcing them to re-implement a stack of runners to do it).

Hopefully that clarifies what you're seeing.

**Comment by bradwilson** [comment]
> You are waiting for testTask to complete, and as soon as it does, you move on to the next test.

This is an interesting design, scheduling something that run that nobody is waiting for. I wonder how/why they landed on this, in the hopes that it would reduce the impact of the reporting? We just did everything inline:

https://github.com/xunit/xunit/blob/6d6b413484e948bffad27efb8326113bf90a7d7f/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs#L250-L270

**Comment by bradwilson** [comment]
Hadn't considered that possibility. I just assumed people were observing ETW events externally with tools like Concurrency Visualizer.

**Comment by mconnew** [comment]
I now have a handle on why I'm seeing failures. The test running parallelization control mechanisms in xunit are not very intuitive. The built in controls are completely insufficient.  

In our CoreWCF.Http.Tests project, we have 73 separate test classes, which by default means we have 73 test collections. Xunit's parallelism constraint is to only allow one test per collection to be run concurrently, but by having 73 test collections, xunit will attempt to run 73 tests concurrently. By default MaxParallelThreads is the number of cpu cores you have, which in Azure Devops is 2. The functionality of MaxParallelThreads is to limit how many threads are used to execute the queued delegates. If a test is synchronous, it will get exclusive usage of the thread until the test has completed.  

Here's a failing scenario. You have an async test which gets started, it does some IO which has a timeout, e.g. an HTTP request with a 30 second timeout (this is exactly the CoreWCF scenario), and it yields the thread through an await as it waits for the response. The next item queued to be executed on the MaxConcurrencySyncContext is a synchronous test which takes 20 seconds to complete. The next item queued to be executed happens to be another synchronous test which also takes 20 seconds to complete. By the time the awaited HTTP request completion continuation gets to execute on a thread, it's timed out and the test fails.  

Fundamentally the approach Xunit takes to parallelism is incompatible with having a mix of async and sync tests running through the same MaxConcurrencySyncContext instance as the sync tests can cause the async tests to be thread starved.  

The solution to this thread starvation has been to set `MaxParallelThreads = -1` as that way the runtime would add new threads when backlogged and the OS would context switch away from a thread running a synchronous test due to being over subscribed on threads to cores which would enable the async tests to make progress and complete.  

The effect that removing `AsyncTestSyncContext` being used by default was to remove an unintended back pressure throttling mechanism on test execution. If I have a class with 20 async methods, they will get executed one at a time. If I have 50 classes (so 50 test collections), Xunit will be concurrently pumping through 50 lists of tests through the test runner. In the class with 20 async methods, when a test completes, this extra delegate from the runtime would also get scheduled to the sync context. The next test from the class wouldn't be started until this extra delegate has finished executing and AsyncTestSyncContext.WaitAsync would complete. When there's a lot of thread pool contention, it might take a bit before this extra delegate gets executed, which causes a delay in starting the next test, which effectively applies back pressure and prevents the test execution from getting overwhelmed. Without AsyncTestSyncContext causing a delay to run the next test where the delay is proportionate to the load, the test execution parallelism overwhelms the CPU and tests start failing.  

I believe there should be a bit of "going back to the drawing board" with the parallelism design in Xunit as it's currently designed to fail if you have a mix of sync and async tests. I think there's some short term changes you can make to improve things: 

Immediately bring back AsyncTestSyncContext for all async tests for it's unintentional side effect as a back pressure mechanism.  

Create an intentional back pressure mechanism. This could be as simple as creating a sync context that queue's everything (like MaxConcurrencySyncContext) and each record/work item placed in the queue records the time it was placed there. As each item is dequeued to run, calculate how long it was waiting in the queue for and have some threshold that sets/resets some flag (which can be awaited such as your AsyncManualResetEvent) which some component which is pumping tests (there's multiple layers which iterate over collections of tests, but I think `XunitTestInvoker.InvokeTestMethodAsync` seems like a good candidate).

If in a thread limited mode (MaxParallelThreads isn't -1), which is the default, separate execution of sync and async tests to use separate `SynchronizationContext` instances so that sync tests can't thread starve async tests.

While I understand the concept of test collections is to prevent them corrupting each others state, e.g. using a shared database instance with test data which gets manipulated, the current implementation of parallelism is based around not corrupting shared state and doesn't take into consideration CPU/thread contention when you have a large test suite. Having some way to limit test collection execution concurrency would get rid of a lot of these problems people are persistently having. I just threw together a prototype which uses a custom TestAssemblyRunner and uses a simple SemaphoreSlim with an initial count of Environment.ProcessorCount along with MaxParallelThreads set to -1 and asynchronously waits on it at the start of RunTestCollectionAsync and my test completion rate has never been faster.

**Comment by bradwilson** [comment]
There is one part of the analysis that isn't clear to me. You mentioned `WaitForCompletionAsync` as being the thing that's alleviating back-pressure, but that only ever ran for tests which are marked as `async void`. Here's the source prior to 2.7.0:

https://github.com/xunit/xunit/blob/b060404f65904e59372ec4e5bfe5260b39db7ded/src/xunit.execution/Sdk/Frameworks/Runners/TestInvoker.cs#L278-L292

You can see that if the test returns a `Task` (of the F# equivalent), then we just await that task and be done. We never call `WaitForCompletionAsync`.

Does this change your back-pressure analysis?

**Comment by bradwilson** [comment]
I have set up a full implementation of your suggestion, and here's an example of the performance difference that I expected to see (`conservative` uses the semaphore, `aggressive` uses the sync context):

```
$ .\src\xunit.console\bin\Release\net462\xunit.console.exe .\test\test.xunit.execution\bin\Release\net462\test.xunit.execution.dll -parallelAlgorithm conservative
xUnit.net Console Runner v2.7.2-pre.7+e59674225e (64-bit .NET Framework 4.6.2, runtime: 4.0.30319.42000)
  Discovering: test.xunit.execution (app domain = off, method display = ClassAndMethod, method display options = None)
  Discovered:  test.xunit.execution (found 545 test cases)
  Starting:    test.xunit.execution (parallel test collections = on [24 threads], stop on fail = off)
  Finished:    test.xunit.execution
=== TEST EXECUTION SUMMARY ===
   test.xunit.execution  Total: 1013, Errors: 0, Failed: 0, Skipped: 0, Time: 15.414s
```

```
$ .\src\xunit.console\bin\Release\net462\xunit.console.exe .\test\test.xunit.execution\bin\Release\net462\test.xunit.execution.dll -parallelAlgorithm aggressive
xUnit.net Console Runner v2.7.2-pre.7+e59674225e (64-bit .NET Framework 4.6.2, runtime: 4.0.30319.42000)
  Discovering: test.xunit.execution (app domain = off, method display = ClassAndMethod, method display options = None)
  Discovered:  test.xunit.execution (found 545 test cases)
  Starting:    test.xunit.execution (parallel test collections = on [24 threads/aggressive], stop on fail = off)
  Finished:    test.xunit.execution
=== TEST EXECUTION SUMMARY ===
   test.xunit.execution  Total: 1013, Errors: 0, Failed: 0, Skipped: 0, Time: 14.300s
```

That's about 8% slower (I choose `test.xunit.execution` because it's the longest running test project that I have in xUnit.net itself).

I believe that I will ship this, but I'm unsure yet which should be the default.

**Comment by bradwilson** [comment]
@mconnew You can now try this out in `2.7.2-pre.8`. The default right now is "conservative", which is your semaphore implementation. It does this _in lieu of_ using the sync context, which I really wanted to try and see what happens, but be aware that because of this, setting a max parallel thread count that is higher than your CPU core count won't do what you think it should do, since it's relying on the built-in ThreadPool-based Task scheduler, which is itself limited to use one thread per CPU core by default (with a very slow uplift, as mentioned previously).

Here's the full commit for your review, if you like: https://github.com/xunit/xunit/commit/0cdf8549f5d6a92c2ce8264c09481d384560fd9b

**Comment by bradwilson** [comment]
And of course doing yet another review of the source I see a bug. 😂

**Comment by bradwilson** [comment]
Okay, seriously, though: `2.7.2-pre.9`.

**Comment by mconnew** [comment]
You might be better off with the following
```c#
if (parallelSemaphore != null)
{
    await parallelSemaphore.WaitAsync(cancellationTokenSource.Token);
}
```
No need to block the thread, and WaitAsync will throw is the CTS is cancelled. There's a lot of things to reply to, I'll try to keep some order to it.

>The problem is that .NET does not give us any system that I'm aware of to measure the time something has actually run vs. the time it spent waiting to run.

This was why I suggested maintaining the Posted time to the synch context with the delegate. Thinking about it more, you don't need a queue, just wrap the posted callback in another callback which records the delay, and wrap the state in another state object which contains the posted time. The wrapping callback unpacks the original state and calls the original callback, but just before calling it, sets something saying how long it waited to execute.  

>wanting an answer in "30 seconds or fail" makes sense in a production system, but the unit testing environment is by definition attempting to be an ultra-high load environment.

One of the tests we now have (this was added as a result of a CVE because it didn't do the right thing) is ensure that CoreWCF cancels/aborts connections after configured timeouts. One variant is our net.tcp transport has a configured ReceiveTimeout, which in production defaults to 10 minutes. In our tests, we set it to a low number like 5 seconds, the client makes a call, then waits for the server to forcefully disconnect it. We call to detect the disconnect is async. We measure how long it took for the server to do so and if not within a certain margin of error (within 2 seconds I think we set), we fail the test. In the WCF project (dotnet/wcf), we have tests from the other side where the client makes a call to the server which purposefully doesn't respond withing a certain amount of time in excess of the clients configured timeout to make sure the client gives up on the request and throws an exception. I'm currently working on the server side implementation of WS-ReliableSessions. This has an ACK mechanism that if the client doesn't ACK a message within a configured period of time, it sends the message again. This all requires a low contention execution environment. We don't require exactness, but waiting much more than a second to execute for a queued work item breaks the test. The only way to validate protocol correctness is to have reliable timing. We also use time limits which if exceeded means something hung/deadlocked and we can't be waiting 20 minutes for xunit to give up to know. Especially if the test is synchronsous and is holding a thread for the entire time. I agree with many tests you don't care how long it takes as long as forward progress is happening, but when testing libraries where a core feature is timing out operations for stability reasons (not replying to things is a classic DoS attack scenario), this approach doesn't work.  

>the built in .NET thread pool has the same default size (1 thread per core) and is super slow to grow (last I checked, 1 thread every 30 seconds)

You're misremembering. At least since .NET 4.0 the algorithm has been keep adding threads until you hit the configured min threads (which is core count by default), then add one more thread every 500ms if there's a backlog of work. Setting it to unlimited works to an extent if you can avoid tests timing out while the thread pool grows.  

> now the OS is the one deciding whether you get to run or not and not us.  

There's a subtle but very important difference. Having extra threads that the OS switches between has better multi-tasking behavior when a thread is blocked. E.g. if waiting for a lock, or doing an IO call with a synchronous api, the effect is the thread immediately yields the rest of it's execution time quota back to the OS for something else to use. When your sync context constrains things to only use 2 threads to dispatch work, a test which is synchronously blocking won't yield the processor to another test to use. Do you remember the old Win3.1 days where multi-threading was cooperative, a program had exclusive use of the CPU until it explicitly yielded back to Windows which could schedule another app to now run, and how a badly behaving program could starve everything else of CPU time, that's what the current max threads parallelism mechanism is mimicking.  

Here's the basic numbers of tests running faster. My dev machine has 20 cores (including HT) and the VS test runner reported the total test time around 26 minutes. My understanding is it adds up the wall clock time that each test took to run, which includes the time await-ing and adds it together. With my change, that number now reports as 9.8 minutes. I haven't checked the actual wall clock execution time for the entire run, but subjectively it seems to complete everything noticeably faster. We have no retry mechanism if a test fails. There is a measurable significant cost to the OS performing a context switch when you are over subscribed on thread count vs core count to the point that even though the async/await Task mechanisms adds a measurable overhead, that overhead is still less than the context switching cost. In real terms, this means a program running 1000 things concurrently on 1000 threads implemented with all IO calls made to win32 sync apis, with lots of blocked time not consuming CPU, that will use more CPU than the equivalent code written using async/await with all the extra overhead of async (overlapped allocation, Task allocation, locals being boxed into the Task state machine and heap allocated etc) because a context switch is so expensive. My set of tests met whatever that threshold was such that running tests more serially and reducing context switches resulted in completing all the tests faster, even if the CPU has more idle time. Over subscribing the CPU with extra threads at some point costs more than serial execution.  

In DevOps, without my semaphore slim, lots of tests were timing out and the entire run on Windows with .Net 8.0 took over 20m minutes to complete with the timeout failures. With my semaphore slim, just shy of 13 minutes to run all the tests.  

You can always over subscribe the test collection parallelism to use more CPU, but that's dependent on the specific set of tests. The same value as the number of cores is a sub optimal safe value which is better than running serially. Hitting the optimal situation would require measuring delegate scheduling to execution latency and dynamically adjusting, and even then an already running test might do something which throws lots of delegates at the sync context and makes that latency spike.  

I did think about providing a way to configure a load factor, basically a double value that is multiplied by the core count and then rounded to the nearest integer. The default load factor could be 1.0, and if you override by specifying 3.0, the Azure Devops CI runs would have a parallelism of 6. While some tests would run a little slower than the could if run serially, the delays wouldn't be sufficiently overwhelming that you would get timeout failures. I like the concept of a load factor rather than a strict number because an attribute doesn't allow dynamic values to be provided. Some devs might have 4 cpu's, others might have 20, and CI often has 2. I don't want that logic in a build task, so a load factor seems like it would be more useful than a count.

RE the back pressure analysis, I agree with you. I had missed that subtlety. I'll have to take another look. The problem I have is everything seems to behave very differently as soon as a debugger is attached.  

I think as long as it's easy to switch between the conservative vs aggressive, I think conservative is the right default. It's better that tests take slightly longer to run than tests fail in a really difficult to analyze way. It is really hard to execute tests locally and mimic the constraints of a Devops CI machine so unless you have just the right domain knowledge about async and sync contexts, discoverability of the cause is really difficult.  

If a parallelism value greater than CPU cores is specified, you could modify minWorkerThreads on the ThreadPool so there's no delay in getting up to that number of threads.

Some useful tidbits, you can reduce the cores available to WSL2 easily by modifying the .wslconfig file. If you setup a hyper-v VM to run Windows with only 2 CPU cores, it will still be too fast, so use sysinternals "CPU Stress" tool to consume CPU cycles in a predictable way (only server OS's allow throttling CPU in Hyper-V as the new scheduler in client SKU's of Windows doesn't support doing that).  

I'll try pulling out my custom implementation and using 2.7.2-pre.9 tomorrow.

**Comment by bradwilson** [comment]
> You might be better off with the following

Good call. 😄 

> The wrapping callback unpacks the original state and calls the original callback, but just before calling it, sets something saying how long it waited to execute.

This might be a fix for Timeout (though honestly, the semaphore-based system probably already fixes that, something I'm about to check), but it wouldn't be able to fix your call into other libraries that are relying on clock time for timeouts.

> One variant is our net.tcp transport has a configured ReceiveTimeout, which in production defaults to 10 minutes. In our tests, we set it to a low number like 5 seconds, the client makes a call, then waits for the server to forcefully disconnect it.

What I have done for things like this in my production code is make the notion of time injectable, rather than relying on something like `DateTimeOffset.Now` that is harder to "change". Then you not only don't have to switch from 10 minutes to 5 seconds (because you can fake the passing of 10 minutes), which means the test runs instantly instead of wasting 5 seconds, and it becomes impervious to the execution environment around it and the state of busyness that's involved. (I've done the same thing for anything which prints/reports/stores the current time as well, so that the current time is always fixed.)

One example: 

https://github.com/xunit/xunit/blob/1fc948f345d0838a52918d66906172f243b38a0f/src/xunit.runner.utility/Sinks/ExecutionSink.cs#L88-L91

https://github.com/xunit/xunit/blob/1fc948f345d0838a52918d66906172f243b38a0f/test/test.xunit.runner.utility/Sinks/ExecutionSinkTests.cs#L107-L121

https://github.com/xunit/xunit/blob/1fc948f345d0838a52918d66906172f243b38a0f/test/test.xunit.runner.utility/Sinks/ExecutionSinkTests.cs#L281-L300

This is the most "complex" I ever see these problems get, because of course I'm waiting for something on another thread/task to wake up and do work, so I am sitting in a relatively tight wait loop (even 25ms is probably overkill). It does, however, require that testability is added to the class in question (something I don't have problems with, but I've seen push back from less "test friendly" developers).

You can even swap out cancellation tokens, using a timeout token in production and a manually managed one in the test environment, if that's the way that timeout is handled.

> Having extra threads that the OS switches between has better multi-tasking behavior when a thread is blocked. E.g. if waiting for a lock, or doing an IO call with a synchronous api, the effect is the thread immediately yields the rest of it's execution time quota back to the OS for something else to use. When your sync context constrains things to only use 2 threads to dispatch work, a test which is synchronously blocking won't yield the processor to another test to use.

I think conflates two things: scheduling of available threads, and determining whether a thread is busy. A synchronous test that's waiting for something that _could've_ been async is most definitely hogging a thread, but that doesn't change with or without our sync context. That's just the way synchronous things are. When the thread pool is limited--whether it's limited by us or by .NET--a blocked thread is a blocked thread. The only difference between us and .NET is the latter has a slow growth system and we don't (which, admittedly, allows them to unblock things--eventually, but not in a timely manner).

> Do you remember the old Win3.1 days where multi-threading was cooperative

All too well, despite therapists telling me to block it out. 😂 

> In DevOps, without my semaphore slim, lots of tests were timing out and the entire run on Windows with .Net 8.0 took over 20m minutes to complete with the timeout failures. With my semaphore slim, just shy of 13 minutes to run all the tests.

It'll be interesting to see if my fix also makes this happen for you (no reason it shouldn't, of course, but that's why we want to verify with a pre-release build before shipping it off to everybody).

> I did think about providing a way to configure a load factor, basically a double value that is multiplied by the core count and then rounded to the nearest integer. The default load factor could be 1.0, and if you override by specifying 3.0, the Azure Devops CI runs would have a parallelism of 6.

From a configuration standpoint, I've add a [core count multiplier syntax](https://xunit.net/docs/configuration-files#maxParallelThreads) for max threads to v3, but it's never been back-ported to v2. A quick and dirty fix would be to update your xunit.runner.json only in CI and add/replace the value for max threads with one that's computed based on the CPU cores.

Back-porting it to the v2 runners is probably not that big a deal; I'll take a look and see.

> If a parallelism value greater than CPU cores is specified, you could modify minWorkerThreads on the ThreadPool so there's no delay in getting up to that number of threads

I have no idea if increasing that after threads are already spun up into the pool works, but if so, then this is also a pretty trivial fix to add (only increase, not decrease).

> The problem I have is everything seems to behave very differently as soon as a debugger is attached.

If I had a dollar every time I said that, I would've been able to retire... at least 6 months before I did. 😂

**Comment by bradwilson** [comment]
This change definitely fixes Timeout issues, and that alone makes `conservative` the for-sure new default.

Purposefully bad tests to illustrate:

```csharp
using System.Threading;
using System.Threading.Tasks;
using Xunit;

namespace Empty;

#pragma warning disable CA1849 // Call async methods when in an async method

public class TestClass1
{
	[Fact(Timeout = 1000)]
	public async Task TestMethod()
	{
		await Task.Yield();
		Thread.Sleep(750);
	}
}

public class TestClass2
{
	[Fact(Timeout = 1000)]
	public async Task TestMethod()
	{
		await Task.Yield();
		Thread.Sleep(750);
	}
}

public class TestClass3
{
	[Fact(Timeout = 1000)]
	public async Task TestMethod()
	{
		await Task.Yield();
		Thread.Sleep(750);
	}
}
```

```
$ xunit.console.exe bin\Debug\net472\empty.dll -maxthreads 1
xUnit.net Console Runner v2.7.2-pre.9+1fc948f345 (64-bit .NET Framework 4.7.2, runtime: 4.0.30319.42000)
  Discovering: empty (app domain = on [shadow copy], method display = ClassAndMethod, method display options = None)
  Discovered:  empty (found 3 test cases)
  Starting:    empty (parallel test collections = on [1 thread], stop on fail = off)
  Finished:    empty
=== TEST EXECUTION SUMMARY ===
   empty  Total: 3, Errors: 0, Failed: 0, Skipped: 0, Time: 2.355s
```

```
$ xunit.console.exe bin\Debug\net472\empty.dll -maxthreads 1 -parallelalgorithm aggressive
xUnit.net Console Runner v2.7.2-pre.9+1fc948f345 (64-bit .NET Framework 4.7.2, runtime: 4.0.30319.42000)
  Discovering: empty (app domain = on [shadow copy], method display = ClassAndMethod, method display options = None)
  Discovered:  empty (found 3 test cases)
  Starting:    empty (parallel test collections = on [1 thread/aggressive], stop on fail = off)
    Empty.TestClass2.TestMethod [FAIL]
      Test execution timed out after 1000 milliseconds
      Stack Trace:
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    Empty.TestClass3.TestMethod [FAIL]
      Test execution timed out after 1000 milliseconds
      Stack Trace:
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
    Empty.TestClass1.TestMethod [FAIL]
      Test execution timed out after 1000 milliseconds
      Stack Trace:
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
        --- End of stack trace from previous location where exception was thrown ---
           at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
           at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
  Finished:    empty
=== TEST EXECUTION SUMMARY ===
   empty  Total: 3, Errors: 0, Failed: 3, Skipped: 0, Time: 2.392s
```

**Comment by bradwilson** [comment]
That said, the potential performance penalty is real. 😢 Which means that `aggressive` is probably here to stay.

```csharp
using System.Threading.Tasks;
using Xunit;

namespace Empty;

public class TestClass1
{
	[Fact(Timeout = 1000)]
	public async Task TestMethod()
	{
		await Task.Delay(750);
	}
}

public class TestClass2
{
	[Fact(Timeout = 1000)]
	public async Task TestMethod()
	{
		await Task.Delay(750);
	}
}

public class TestClass3
{
	[Fact(Timeout = 1000)]
	public async Task TestMethod()
	{
		await Task.Delay(750);
	}
}
```

conservative: `Total: 3, Errors: 0, Failed: 0, Skipped: 0, Time: 2.342s`
aggressive: `Total: 3, Errors: 0, Failed: 0, Skipped: 0, Time: 0.831s`

**Comment by bradwilson** [comment]
If you haven't started your testing yet @mconnew, I will shortly have a new build which enables the multipler thread count syntax (e.g., `2.0x` == 2 * the number of CPU threads) as well as ensuring that we raise the minimum thread pool size if necessary to accommodate max parallel threads > max CPU threads.

**Comment by bradwilson** [comment]
Available now in `2.7.2-pre.13`

**Comment by mconnew** [comment]
I previously had this attribute:
```c#
[assembly: Xunit.CollectionBehavior(MaxParallelThreads = -1)]
```
I commented that out and configured with xunit.runner.json instead. Configuring with either conservative or aggressive with maxParallelThreads as "2.0x" runs the tests with similar total time, with aggressive having slightly lower total test execution time (9.1min vs 9.6min). The odd thing is when leaving the attribute in place, the test execution doesn't run anything. This is what's in the Tests output in VS:
```
========== Starting test run ==========
[xUnit.net 00:00:00.00] xUnit.net VSTest Adapter v2.5.7+8f2703126a (64-bit .NET 6.0.29)
[xUnit.net 00:00:00.08] CoreWCF.Http.Tests: Multiple test collections declared with name 'HttpSysTestCollection': WSHttp.SimpleWSHTTPTest, CoreWCF.Http.Tests.Helpers.HttpSysTestCollectionDefinition
[xUnit.net 00:00:00.15]   Starting:    CoreWCF.Http.Tests (parallel test collections = on [20 threads], stop on fail = off)
[xUnit.net 00:00:00.17]   Finished:    CoreWCF.Http.Tests
========== Test run finished: 0 Tests (0 Passed, 0 Failed, 0 Skipped) run in 180 ms ==========
```
The content of my xunit.runner.json file when this happened was this:
```json
{
    "$schema": "https://xunit.net/schema/current/xunit.runner.schema.json",
    "diagnosticMessages": true
}
```
I don't know how much time I'll get before Monday to dig into this more, but thought I would raise this as something potentially being broken. I also didn't get time to test it in a really CPU constrained environment yet.

**Comment by bradwilson** [comment]
> I would raise this as something potentially being broken

I think there are two issues.

First is the thread count display. It's a weird situation. The "meta" runners (the console runner, the VS Test adapter, etc.) don't know anything about the internal workings of how the tests run, so they cannot "see" `CollectionBehavior` in order to update their display. They can only show what "advice" they've given the execution system (via the execution options), which of course the execution engine is free to ignore. So the display will be wrong when relying on the attribute. It will be correct when relying on command line switches or configuration files, because those are read by the "meta" runners to provide that advice to the execution engine. So, yes it's a bug, but we have opted not to fix it because of the changes that would be required to do so.

Second is why no test ran. That one I'm able to repro locally. And I get this dreaded problem when running it on the console:

```
$ xunit.console.exe .\bin\Debug\net462\empty.dll
xUnit.net Console Runner v2.7.2-pre.13+3722e54234 (64-bit .NET Framework 4.6.2, runtime: 4.0.30319.42000)
  Discovering: empty (app domain = on [shadow copy], method display = ClassAndMethod, method display options = None)
  Discovered:  empty (found 1 test case)
  Starting:    empty (parallel test collections = on [24 threads], stop on fail = off)
  Finished:    empty
System.AppDomainUnloadedException: Attempted to access an unloaded AppDomain.
```

That's not gonna be fun.

**Comment by bradwilson** [comment]
Oh, hey, turn off app domains and voila:

```
System.ArgumentOutOfRangeException: The initialCount argument must be non-negative and less than or equal to the maximumCount.
Parameter name: initialCount
Actual value was -1.
   at System.Threading.SemaphoreSlim..ctor(Int32 initialCount, Int32 maxCount)
```

That's a bug I already noticed and fixed! Just need to push it and get a nupkg I can test with.

**Comment by bradwilson** [comment]
`2.7.2-pre.17` has the fix for the `[assembly: Xunit.CollectionBehavior(MaxParallelThreads = -1)]` issue.

**Comment by bradwilson** [comment]
@mconnew I'm working on the documentation pages for this, and I wonder if you have any comments/corrections/additions. I'll probably include this text in the release notes as well (or at the very least, a link over to this doc page).

----------

<h2>Algorithms</h2>

<p>
  As of v2 Test Framework 2.8.0, we've changed the default parallelism algorithm,
  with the ability to fall back the original. The two algorithms are called
  <code>conservative</code> (the new default) and <code>aggressive</code> (the
  original algorithm). They have the following attributes:
</p>

<code>conservative</code>

<blockquote>
  <p>
    This will only <em>start</em> as many tests as your max parallel threads
    setting. The system will wait for a test to finish before starting another
    test. This allows more accurate timing of the running of tests, which
    allows <code>Timeout</code> on <code>[Fact]</code> to work as expected even
    when running tests in parallel. It also creates less pressure on the task
    system, which has occasionally caused deadlocks in complex projects. The
    downside is that projects with lots of tests which spend a significant amount
    of time waiting for async operations to complete will under-utilize the
    maximum CPU potential, so this new algorithm may cause your overall test
    run to be slower.
  </p>
</blockquote>

<code>aggressive</code>

<blockquote>
  <p>
    This is the original parallelism algorithm, which starts as many tests as
    possible, regardless of your max parallel threads setting, and uses a
    <a href="https://learn.microsoft.com/en-us/dotnet/api/system.threading.synchronizationcontext"><code>SynchronizationContext</code></a>
    to limit the number of things that are running at any given time. Since
    tests in this system which encounter async awaits and put back into a pool
    to compete against all potential running tests, they may wait longer to resume
    which causes the inaccuracy of timing that makes <code>Timeout</code> problematic.
    On the flip side, projects with lots of tests which spend a significant amount
    of time waiting for async operations will make better use of CPU resources,
    assuming there are more tests ready to run, so this old algorithm may
    finish running your tests quicker than the new algorithm.
  </p>
</blockquote>

<p>With Runners 2.8.0+, you can specify overrides for the parallelism algorithm.</p>

<ul>
  <li>You can specify a value in your <a href="configuration-files">configuration file</a></li>
  <li>The console runner can specify the value via the <code>-parallelalgorithm</code> switch</li>
  <li>The MSBuild runner can specify the value via the <code>ParallelAlgorithm</code> property</li>
  <li>The Visual Studio runner can specify the value via <a href="runsettings">RunSettings</a></li>
</ul>

**Comment by mconnew** [comment]
I've finished local testing on a 2 core CPU constrained VM and nothing is timing out, but it's also not timing out when setting aggressive mode. This is the contents of my xunit.runner.json file"
```json
{
    "$schema": "https://xunit.net/schema/current/xunit.runner.schema.json",
    "diagnosticMessages": true,
    "maxParallelThreads": "-1",
    "parallelAlgorithm": "aggressive"
}
```
I'm not sure if I configured things correctly as this is part of the output:
```
Starting:    CoreWCF.Http.Tests (parallel test collections = on [2 threads], stop on fail = off)
```
Based on commit https://github.com/xunit/xunit/commit/6790b481b2cdf5a0502e92e86336b42c9052851b I would expect it to say `[2 threads/aggressive]` in the output. I also tried with all lower case: `"parallelalgorithm"` too. I'm am using xunit 2.7.2-pre.17. As for the text, a couple of grammar things:

Missing the word "to":

As of v2 Test Framework 2.8.0, we've changed the default parallelism algorithm, with the ability to fall back to the original. 

Replace "and" -> "are":

Since tests in this system which encounter async awaits are put back into a pool to compete against all potential running tests,

**Comment by bradwilson** [comment]
> Based on commit https://github.com/xunit/xunit/commit/6790b481b2cdf5a0502e92e86336b42c9052851b I would expect it to say `[2 threads/aggressive]` in the output.

Oops, I didn't release an updated version of `xunit.runner.visualstudio` that would read that setting, so that's my bad. I'll do that right now. Sorry about that. It's available now in `2.5.9-pre.2` (which will eventually become `2.8.0` to make documentation simpler).

Thanks for the doc review!

**Comment by mconnew** [comment]
Please don't forget to update the json schema, I keep spelling aggressive wrong (I keep using only one g) and then wonder why it's not doing what I'm wanting. Updating the json schema would help with that.  

I can confirm using the new vs runner, on a 2 core VM with 60% CPU consumption to increase contention, -1/aggressive has timeouts and something odd was happening where it got up to over 25 minutes of VS reported test time and only had executed 19 tests and had been running for over 5 minutes. I think this is what happens in our CI runs. To give some timing context, when using 2.0x/conservative, vs reports 9.6min summed test execution duration and the output window says 2.5min wall clock time to complete, so a major improvement for my scenario. With 2.0x it correctly reported using 4 threads in the output, which is correct for this VM. 

Running 2.0x/conservative on my main dev box (20 cores) took 9.5min summed test execution time and 1 min wall clock time. It seems like 9.5 mins is the total time all the tests really need to execute. Running -1/aggressive caused test timeouts where the original implementation never had timeouts on my dev machine, but before vs was reporting more than 12 mins total test execution time so there must have been some contention before as that's greater than 9.5 mins. The contention has increased as now vs is saying total duration is 201.4 minutes, and the test output said wall clock time was 7.5 minutes.   

When specifying -1, what does that mean? The test output says 2 threads (VM has 2 cores) when specifying -1 in the config file.

Overall I think these changes are great. If you have no timeout issues and want things to run as fast as possible, I think the changes will make things faster. I suspect that's due to modifying the min threads to remove warmup time. And for those of us who get timeouts with thread contention, the new conservative mode works great.

Do you have an idea of when you would release this RTW? I have a pending PR with my custom XUnit test runner that I'd like to drop, and I want to rework our running profiles to use xunit.runner.json instead of attributes, and so I can use the multiplier format. But I would rather not merge an addition to my nuget.config file.

**Comment by bradwilson** [comment]
> Updating the json schema would help with that.

Yep, that'll be part of the 2.8.0 release. 😄

> it got up to over 25 minutes of VS reported test time and only had executed 19 tests and had been running for over 5 minutes

> when using 2.0x/conservative, vs reports 9.6min summed test execution duration and the output window says 2.5min wall clock time to complete

This is a UI choice here from Test Explorer: rather than showing clock time, they show the sum of all the execution time that has completed thus far. Given enough CPU cores and enough parallelism, you'll see this number rise dramatically faster than clock time. They could opt to show both, in my opinion, but they choose not to.

> When specifying -1, what does that mean?

There are effectively three parallelism algorithms now when there were two, and the third is "don't constrain parallelism". That means we don't use the semaphore, and we don't use our sync context. Instead, we rely on the .NET Thread Pool.

> The test output says 2 threads (VM has 2 cores) when specifying -1 in the config file.

What I'd expect to see depends on _how_ you specify -1. As explained earlier, specifying this via the `CollectionBehavior` attribute is not visible to the runner, so it'll display whatever it thinks should be used. If you set it via runsettings or `xunit.runner.json`, then it should show "unlimited".

Unlimited via JSON:

![image](https://github.com/xunit/xunit/assets/16944/8378ce83-16c1-441f-9bc7-2ba43c82d6b9)

Unlimited via attribute (I have a 24 thread CPU):

![image](https://github.com/xunit/xunit/assets/16944/4f34a85e-6781-4008-8537-4165055c98d5)

Seeing 2 in your output is not surprising if you're using the attribute, given that your VM has two CPUs.

That said, this does bring up a minor UI bug: I actually print `[unlimited threads/aggressive]` when aggressive is irrelevant. I'll fix that before 2.8.0.

> Do you have an idea of when you would release this RTW?

I'm waiting for a combination of things, one of which was your confirmation that we've nailed the problem. Then I need to fix up all the ancillary things like docs, schema, etc. Unfortunately I've been sick since Friday so I haven't had much time to spend on any of this but I'm hoping that I'll be able to get to it this week. AFAIK there's nothing else waiting to go in.

**Comment by mconnew** [comment]
I was specifying -1 via the json method, but I do have one Collection explicitly defined, but it's just declaring a collection, there's no config on it. I'm not too concerned as I'm specifying 2.0x now which I believe is close to optimal for my scenario. I've kicked off an Azure Devops CI build using the pre release xunit to do one final validation. The only thing I want to change before merging is switching to the final build from nuget when it's available.  

I appreciate you implementing semaphore mechanism at the collection running level, I think it's going to get rid of a lot of failure noise for many people.

**Comment by mconnew** [comment]
Devops run was successful.

**Comment by bradwilson** [comment]
Finally complete! 🎉

Available in v2: `2.7.2-pre.21`
Available in v3: `0.1.1-pre.407`
Available in xunit.runner.visualstudio `2.5.9-pre.2`

The v2 Core Framework and `xunit.runner.visualstudio` versions will be numbered 2.8.0 upon release. Hoping to get to that in the next few days.

**Comment by bradwilson** [comment]
@ajeglin I know there was some Moq confusion in the mix, but it would be good to try the pre-release builds and see if your issues are resolved. The new default parallel algorithm should create a lot less contention in the TPL.

@BEagle1984 I don't know what the status of your deadlocks are, but the new algorithm _should_ make it more difficult to deadlock things, since we won't launch more tests than we know you have threads, so it's definitely worth trying out the pre-release build for you, too.

**Comment by bradwilson** [comment]
FYI this landed in release today via 2.8.0: https://dotnet.social/@xunit/112344769563402988

> We just shipped v2 Core Framework 2.8.0, Analyzers 1.13.0, and Visual Studio adapter 2.8.0. The primary purpose of this release is a new parallelism algorithm that should make test timing more reliable, and make thread deadlocks in your tests less likely.
>
> Read all about the new algorithm here: https://xunit.net/docs/running-tests-in-parallel#algorithms
>
> Enjoy!
>
> Release notes:
>
> https://xunit.net/releases/v2/2.8.0
> https://xunit.net/releases/analyzers/1.13.0
> https://xunit.net/releases/visualstudio/2.8.0
>
> #dotnet #csharp #fsharp #unittest #tdd

**Comment by ajeglin** [comment]
> FYI this landed in release today via 2.8.0: https://dotnet.social/@xunit/112344769563402988
> 
> > We just shipped v2 Core Framework 2.8.0, Analyzers 1.13.0, and Visual Studio adapter 2.8.0. The primary purpose of this release is a new parallelism algorithm that should make test timing more reliable, and make thread deadlocks in your tests less likely.
> > Read all about the new algorithm here: https://xunit.net/docs/running-tests-in-parallel#algorithms
> > Enjoy!
> > Release notes:
> > https://xunit.net/releases/v2/2.8.0
> > https://xunit.net/releases/analyzers/1.13.0
> > https://xunit.net/releases/visualstudio/2.8.0
> > #dotnet #csharp #fsharp #unittest #tdd

@bradwilson thank you - I can confirm that our tests no longer hang on version 2.8.0. Thanks so much for all your efforts getting to the bottom of this :)

**Comment by mconnew** [comment]
@bradwilson, thanks again for making this change. I don't see any mention in the docs about the `2.0x` style max threads value in config files. Is this a doc omission or did that get pulled in the final version?

**Comment by bradwilson** [comment]
I did also mention it in the release notes, with links to the docs (under "Runner Utility" in the main release notes, and "Features and Fixes" in the VS Runner release notes).

**Comment by bradwilson** [comment]
@mconnew Thanks, I've updated https://xunit.net/docs/running-tests-in-parallel#parallelism-in-runners to show the multiplier syntax for both Console and MSBuild runners.
